docs/csharp/tutorials/pattern-matching.md

```diff
@@ -0,0 +1,325 @@
+
+## Add occupancy pricing
+
+The toll authority wants to encourage vehicles to travel at maximum capacity. They've decided to charge more when vehicles have fewer passengers, and encourage full vehicles by offering lower pricing:
+
+- Cars and taxis with no passengers pay an extra $0.50.
+- Cars and taxis with two passengers get a 0.50 discount.
+- Cars and taxis with three or more passengers get a $1.00 discount.
+- Buses that are less than 50% full pay an extra $2.00.
+- Buses that are more than 90% full get a $1.00 discount.
+
+These rules can be implemented using the **property pattern** in the same switch expression. The property pattern examines properties of the object once the type has been determined.  The single case for a `Car` expands to four different cases:
+
+```csharp
+Car { Passengers: 0} => 2.00m + 0.50m,
+Car { Passengers: 1 } => 2.0m,
+Car { Passengers: 2} => 2.0m - 0.50m,
+Car c when c.Passengers > 2 => 2.00m - 1.0m,
+```
+
+The first three cases test the type as a `Car`, then check the value of the `Passengers` property. If both match, that expression is evaluated and returned. The final clause shows the `when` clause for a property pattern. You use the `when` clause to test conditions other than equality on a property. In the preceding example, the `when` clause tests to see that there are more than 2 passengers in the car. Strictly speaking, it's not necessary in this example.
+
+You would also expand the cases for taxis in a similar manner:
+
+```csharp
+Taxi { Fares: 0} => 3.50m + 1.00m,
+Taxi { Fares: 1 } => 3.50m,
+Taxi { Fares: 2} => 3.50m - 0.50m,
+Taxi t => 3.50m - 1.00m,
+```
+
+In the preceding example, the `when` clause was omitted on the final case.
+
+Next, implement the occupancy rules by expanding the cases for buses, as shown in the following exmaple:
+
+```csharp
+Bus b when ((double)b.Riders / (double)b.Capacity) < 0.50 => 5.00m + 2.00m,
+Bus b when ((double)b.Riders / (double)b.Capacity) > 0.90 => 5.00m - 1.00m, 
+Bus b => 5.00m,
+```
+
+The toll authority isn't concerned with the number of passengers in the delivery trucks. Instead, they charge more based on the weight class of the trucks. Trucks over 5000 lbs are charged an extra $5.00. Light trucks, under 3000 lbs are given a $2.00 discount.  That rule is implemented with the following code:
+
+```csharp
+DeliveryTruck t when (t.GrossWeightClass > 5000) => 10.00m + 5.00m,
+DeliveryTruck t when (t.GrossWeightClass < 3000) => 10.00m - 2.00m,
+DeliveryTruck t => 10.00m,
+```
+
+## Recursive patterns
+
+You can make this code less repetitive by using **recursive patterns**. The `Car` and `Taxi` both have four different arms in the preceding examples. In both cases, you can create a type pattern that feeds into a property pattern. This technique is shown in the following code:
```

Comment:
nested switch expressions?

Reply:
All the docs from the product team refer to this as "recursive patterns".

We should be consistent.